lib-src/libcs/octarg.3
.\" COPYRIGHT NOTICE
.\" Copyright (c) 1994 Carnegie Mellon University
.\" All Rights Reserved.
.\" 
.\" See <cmu_copyright.h> for use and distribution information.
.\" 
.\" 
.\" HISTORY
.\" $Log: octarg.3,v $
.\" Revision 4.1  1997/01/08 21:53:59  rvb
.\" r = 4.1; fix $ HEADERS
.\"
.\" Revision 1.1  1996/11/22 19:19:13  braam
.\" First Checkin (pre-release)
.\"
Revision 1.1  96/06/03  19:00:52  satya
Initial revision

.\" Revision 1.2  1995/10/18  14:58:46  moore
.\" 	Created libcs man(3) pages from mach entries
.\" 	[1995/10/18  14:55:17  moore]
.\"
.\" $EndLog$
.\" Copyright (c) 1990 Carnegie Mellon University
.\" All Rights Reserved.
.\" 
.\" Permission to use, copy, modify and distribute this software and its
.\" documentation is hereby granted, provided that both the copyright
.\" notice and this permission notice appear in all copies of the
.\" software, derivative works or modified versions, and any portions
.\" thereof, and that both notices appear in supporting documentation.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND CARNEGIE MELLON UNIVERSITY
.\" DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS.  IN NO EVENT
.\" SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE FOR ANY SPECIAL, DIRECT,
.\" INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
.\" RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
.\" CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
.\" CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\" Users of this software agree to return to Carnegie Mellon any
.\" improvements or extensions that they make and grant Carnegie the
.\" rights to redistribute these changes.
.\"
.\" Export of this software is permitted only after complying with the
.\" regulations of the U.S. Deptartment of Commerce relating to the
.\" Export of Technical Data.
.\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\" HISTORY
.\" $Log: octarg.3,v $
.\" Revision 4.1  1997/01/08 21:53:59  rvb
.\" r = 4.1; fix $ HEADERS
.\"
.\" Revision 1.1  1996/11/22 19:19:13  braam
.\" First Checkin (pre-release)
.\"
Revision 1.1  96/06/03  19:00:52  satya
Initial revision

.\" Revision 1.2  1995/10/18  14:58:46  moore
.\" 	Created libcs man(3) pages from mach entries
.\" 	[1995/10/18  14:55:17  moore]
.\"
.\" Revision 1.1.1.2  1995/10/18  14:55:17  moore
.\" 	Created libcs man(3) pages from mach entries
.\"
.\" Revision 1.2  90/12/12  15:54:35  mja
.\" 	Add copyright/disclaimer for distribution.
.\" 
.\" 13-Nov-86  Andi Swimmer (andi) at Carnegie-Mellon University
.\" 	Revised for 4.3.
.\" 
.\" 05-Dec-79  Steven Shafer (sas) at Carnegie-Mellon University
.\" 	Created.
.\" 
.TH OCTARG 3 12/5/79
.CM 1
.SH "NAME"
octarg, hexarg \- parse octal or hexadecimal argument or ask user
.SH "SYNOPSIS"
.B
unsigned int octarg (ptr,brk,prompt,min,max,defalt);
.br
.B
unsigned int hexarg (ptr,brk,prompt,min,max,defalt);
.sp
.B
char **ptr,*brk,*prompt;
.br
.B
unsigned int min,max,defalt;
.SH "DESCRIPTION"
.I
octarg
and
.I
hexarg
attemp to parse an argument from a string, passing the
string pointer
.I
ptr
and the break character set
.I
brk
to the
.IR nxtarg (3)
routine.
If there is an argument parsed, and it is a legal
octal or hexadecimal value between
.I
min
and
.I
max,
then this value is returned by
.I
octarg
or
.I
hexarg.
If there is no argument, or it is not a legal value, then
an error message is printed and the remaining parameters
are passed into
.IR getoct (3)
or
.IR gethex (3),
whose value will be returned by
.I
octarg
or
.I
hexarg.
.sp
A legal octal value is a sequence of octal digits.
A legal hexadecimal
value is a sequence of digits "0" through "9", "a" through "f", or
"A" through "F", and optionally preceded by the (ignored) prefix
"0x" or "0X".
.sp
.I
octarg
and
.I
hexarg
are especially useful for parsing octal and hexadecimal values from
argument lists in programs using the command interpreter,
.IR ci (3).
.SH "LIBRARY NOTES"
You must link against /usr/local/lib/libcs.a to define this routine.
.SH "SEE ALSO"
.IR nxtarg (3), 
.IR getoct (3), 
.IR gethex (3), 
.IR atoo (3), 
.IR atoh (3), 
.IR ci (3)


NEWS
Incompatible RVM update (Coda client)
-------------------------------------

2008/02/06 RVM version 39
    Move store identifier from per-volume to global rvg struct.

2005/06/07 RVM version 38
    Added realm::realmid (as we can't use &realm_object on 64-bit systems)

2005/06/16 RVM version 37
    Removed fsdb::device and fsobj::inode


lib-src/libcs/octarg.c
#ifndef _BLURB_
#define _BLURB_
/*
 * This code was originally part of the CMU SCS library "libcs".
 * A branch of that code was taken in May 1996, to produce this
 * standalone version of the library for use in Coda and Odyssey
 * distribution.  The copyright and terms of distribution are
 * unchanged, as reproduced below.
 *
 * Copyright (c) 1990-96 Carnegie Mellon University
 * All Rights Reserved.
 * 
 * Permission to use, copy, modify and distribute this software and its
 * documentation is hereby granted, provided that both the copyright
 * notice and this permission notice appear in all copies of the
 * software, derivative works or modified versions, and any portions
 * thereof, and that both notices appear in supporting documentation.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND CARNEGIE MELLON UNIVERSITY
 * DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS.  IN NO EVENT
 * SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE FOR ANY SPECIAL, DIRECT,
 * INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
 * RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
 * CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * Users of this software agree to return to Carnegie Mellon any
 * improvements or extensions that they make and grant Carnegie the
 * rights to redistribute these changes.
 *
 * Export of this software is permitted only after complying with the
 * regulations of the U.S. Deptartment of Commerce relating to the
 * Export of Technical Data.
 */

static char *rcsid = "$Header: /afs/cs.cmu.edu/project/coda-braam/src/coda-4.0.1/RCSLINK/./lib-src/libcs/octarg.c,v 1.1 1996/11/22 19:19:41 braam Exp $";
#endif /*_BLURB_*/

/*  octarg  --  parse octal integer argument
 *
 *  Usage:  i = octarg (ptr,brk,prompt,min,max,default)
 *    unsigned int i,min,max,default;
 *    char **ptr,*brk,*prompt;
 *
 *  Will attempt to parse an argument from the string pointed to
 *  by "ptr", incrementing ptr to point to the next arg.  If
 *  an arg is found, it is converted into an octal integer.  If there is
 *  no arg or the value of the arg is not within the range
 *  [min..max], then "getoct" is called to ask the user for an
 *  octal integer value.
 *  "Brk" is the list of characters which may terminate an argument;
 *  if 0, then " " is used.
 *
 */

#include <stdio.h>
#include <ctype.h>
#include <string.h>

#include "libcs.h"

unsigned int octarg (ptr,brk,prompt,min,max,defalt)
char **ptr;
const char *brk,*prompt;
unsigned int min,max,defalt;
{
	register unsigned int i = 0;
	register char *arg,*p = NULL;

	arg = nxtarg (ptr,brk);
	fflush (stdout);

	if (*arg != '\0') {		/* if there was an arg */
		for (p=arg; *p && (*p >= '0' && *p <= '7'); p++) ;
		if (*p) {
			if (strcmp(arg,"?") != 0)  fprintf (stderr,"%s not octal number.  ",arg);
		} 
		else {
			i = atoo (arg);
			if (i<min || i>max) {
				fprintf (stderr,"%s%o out of range.  ",
					(i ? "0" : ""),i);
			}
		}
	}

	if (*arg == '\0' || (p && *p != '\0') || i<min || i>max) {
		i = getoct (prompt,min,max,defalt);
	}

	return (i);
}
